modules/web/src/com/haulmont/addon/imap/web/imapfolderevent/ImapFolderEventEdit.java

```diff
@@ -0,0 +1,258 @@
+
+import com.haulmont.addon.imap.entity.*;
+import com.haulmont.addon.imap.service.ImapService;
+import com.haulmont.cuba.core.global.AppBeans;
+import com.haulmont.cuba.core.global.Metadata;
+import com.haulmont.cuba.gui.components.*;
+import com.haulmont.cuba.gui.data.CollectionDatasource;
+import com.haulmont.cuba.gui.data.Datasource;
+import com.haulmont.cuba.gui.xml.layout.ComponentsFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import java.util.*;
+import java.util.stream.Collectors;
+
+@SuppressWarnings("CdiInjectionPointsInspection")
+public class ImapFolderEventEdit extends AbstractEditor<ImapFolderEvent> {
+
+    private final static Logger log = LoggerFactory.getLogger(ImapFolderEventEdit.class);
+
+    @Inject
+    protected Datasource<ImapFolderEvent> imapFolderEventDs;
+
+    @Inject
+    private CollectionDatasource<ImapEventHandler, UUID> handlersDs;
+
+    @Inject
+    private Table<ImapEventHandler> handlersTable;
+
+    @Inject
+    private Button addHandlerBtn;
+
+    @Inject
+    private Button upHandlerBtn;
+
+    @Inject
+    private Button downHandlerBtn;
+
+    @Inject
+    private ImapService service;
+
+    @Override
+    protected void postInit() {
+        super.postInit();
+
+        ImapFolderEvent folderEvent = getItem();
+
+        ImapEventType eventType = folderEvent.getEvent();
+        Map<String, List<String>> availableBeans = eventType != null
+                ? service.getAvailableBeans(eventType.getEventClass()) : Collections.emptyMap();
+        List<String> beanNames = new ArrayList<>(availableBeans.keySet());
+        long maxHandlersCount = availableBeans.values().stream().mapToLong(Collection::size).sum();
+
+        removeMissedHandlers(availableBeans);
+        enableAddButton(maxHandlersCount);
+
+        if (availableBeans.isEmpty()) {
+            handlersTable.getParent().setVisible(false);
+            return;
+        }
+
+        Map<ImapEventHandler, LookupField> handlerMethodLookups = new HashMap<>();
+        if (folderEvent.getEventHandlers() != null) {
+            folderEvent.getEventHandlers().forEach(eventHandler ->
+                    handlerMethodLookups.put(eventHandler, makeBeanMethodLookup(availableBeans, eventHandler))
+            );
+        }
+
+        addHandlersCollectionChangeListeners(availableBeans, maxHandlersCount, handlerMethodLookups);
+        setHandlerChangeListeners(availableBeans, handlerMethodLookups);
+
+        generateColumns(availableBeans, beanNames, handlerMethodLookups);
+
+        addCloseWithCommitListener(() -> {
+
+            @SuppressWarnings("unchecked")
+            Datasource<ImapFolderEvent> parentDs = getParentDs();
+            if (parentDs != null && parentDs.getItem() != null) {
+                parentDs.getItem().setEventHandlers(folderEvent.getEventHandlers());
+                for (int i = 0; i < folderEvent.getEventHandlers().size(); i++) {
```

Comment:
What is the reason not to use "for each" loop?

Reply:
To set handlingOrder of handler directly: folderEvent.getEventHandlers().get(i).*setHandlingOrder(i)*